src/compiler/tsbuild.ts

```diff
@@ -273,6 +274,26 @@ namespace ts {
     export interface SolutionBuilderWithWatchHost<T extends BuilderProgram> extends SolutionBuilderHostBase<T>, WatchHost {
     }
 
+    /*@internal*/
+    export type BuildOrder = readonly ResolvedConfigFileName[];

```

Comment:
I think this would be easier to read as
```suggestion
    export type BuildOrder = { files: readonly ResolvedConfigFileName[]; circularDiagnostics?: readonly Diagnostic[] }
```

followed by deleting `CircularBuildOrder` and `AnyBuildOrder` and then isCircularBuildOrder could check for the presence of `circularDiagnostics` instead of `buildOrder` (and could just return `boolean`).

Reply:
I don't want to wrap things in object just with array. I think I like this better with order being Array and CirucularBuildOrder diagnostics only when circular build order. This typing also helps in making sure we are handling circular build order in any place build order plays the role.

Reply:
Fair enough.